src/Polly.Core/Simmy/MonkeyStrategy.cs

```diff
@@ -0,0 +1,78 @@
+    /// </summary>
+    /// <param name="options">The chaos strategy options.</param>
+    protected MonkeyStrategy(MonkeyStrategyOptions options)
+    {
+        Guard.NotNull(options);
+        Guard.NotNull(options.Randomizer);
+
+        if (options.InjectionRate is null && options.InjectionRateGenerator is null)
+        {
+            throw new ArgumentNullException(nameof(options.InjectionRate), "Either InjectionRate or InjectionRateGenerator is required.");
+        }
+
+        if (options.Enabled is null && options.EnabledGenerator is null)
+        {
+            throw new ArgumentNullException(nameof(options.Enabled), "Either Enabled or EnabledGenerator is required.");
+        }
+
+        _randomizer = options.Randomizer;
+        InjectionRateGenerator = options.InjectionRate.HasValue ? (_) => new(options.InjectionRate.Value) : options.InjectionRateGenerator;
+        EnabledGenerator = options.Enabled.HasValue ? (_) => new(options.Enabled.Value) : options.EnabledGenerator;
+    }
+
+    /// <summary>
+    /// Gets the injection rate for a given execution, which the value should be between [0, 1].
+    /// </summary>
+    public Func<ResilienceContext, ValueTask<double>> InjectionRateGenerator { get; }
+
+    /// <summary>
+    /// Gets a value that indicates whether or not the chaos strategy is enabled for a given execution.
+    /// </summary>
+    public Func<ResilienceContext, ValueTask<bool>> EnabledGenerator { get; }
+
+    /// <summary>
+    /// Determines whether or not the chaos strategy should be injected based on the injection rate and enabled flag.
+    /// </summary>
+    /// <param name="context">The <see cref="ResilienceContext"/> instance.</param>
+    /// <returns>A boolean value that indicates whether or not the chaos strategy should be injected.</returns>
+    /// <remarks>Use this method before injecting any chaos strategy to evaluate whether a given chaos strategy needs to be injected during the execution.</remarks>
+    public async ValueTask<bool> ShouldInject(ResilienceContext context)
+    {
+        Guard.NotNull(context);
+
+        // to prevent executing config delegates if token was signaled before to start.
+        context.CancellationToken.ThrowIfCancellationRequested();
+
+        if (!await EnabledGenerator(context).ConfigureAwait(context.ContinueOnCapturedContext))
+        {
+            return false;
+        }
+
+        // to prevent executing InjectionRate config delegate if token was signaled on Enabled configuration delegate.
+        context.CancellationToken.ThrowIfCancellationRequested();
+
+        double injectionThreshold = await InjectionRateGenerator(context).ConfigureAwait(context.ContinueOnCapturedContext);
+
+        // to prevent executing further config delegates if token was signaled on InjectionRate configuration delegate.
+        context.CancellationToken.ThrowIfCancellationRequested();
+
+        injectionThreshold.EnsureInjectionThreshold();
```

Comment:
I would suggest not throwing at runtime in this case. Instead, maybe just coerce the value or just return false on invalid one?

Reply:
Curious to know your thoughts on why it's not good to throw at runtime and instead use default values? I like both suggestions tho

Reply:
The idea is that if strategy can replace the value with reasonable default i.e.

- If user generator returns value less than 0, we can just say it's 0 and it won't have any effect on the execution.
- If user generator returns value greater than 0, we can just say it's 1 and it won't have any effect on the execution.

This really applies if we can provide those defaults, sometimes we cannot assume anything and it's better to throw.

I have applied the same principle in the timeout strategy:
https://github.com/App-vNext/Polly/blob/fc58ddcacafd1c8e55c31ba7ad8fef3406bb5973/src/Polly.Core/Timeout/TimeoutResilienceStrategy.cs#L36

(avoid throwing if invalid timeout is provided and just ignore the value)